engine/classes/Elgg/Router.php

```diff
@@ -70,42 +74,50 @@ public function route(\Elgg\Http\Request $request) {
 			$this->timer->begin(['build page']);
 		}
 
+		ob_start();
+
 		$result = $this->hooks->trigger('route', $identifier, $old, $old);
 		if ($result === false) {
-			return true;
-		}
-
-		if ($result !== $old) {
+			$output = ob_get_clean();
+			$response = elgg_ok_response($output);
+		} else if ($result !== $old) {
 			_elgg_services()->logger->warn('Use the route:rewrite hook to modify routes.');
-		}
+			$query = parse_url(current_page_url(), PHP_URL_QUERY);
+			$segments = (array) elgg_extract('segments', $result, []);
+			$identifier = elgg_extract('identifier', $result);
+			array_unshift($segments, $identifier);
+			_elgg_services()->responseFactory->redirect(implode('/', $segments) . '?' . $query);
```

Comment:
Sounds fine. We can just something to docs that plugins still using route to rewrite routes need to also use the original route for filtering response values.